internal/mesh/internal/mappers/sidecarproxymapper/destinations_mapper.go

```diff
@@ -0,0 +1,67 @@
+package sidecarproxymapper
+
+import (
+	"context"
+
+	"github.com/hashicorp/consul/internal/catalog"
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/mesh/internal/types"
+	"github.com/hashicorp/consul/internal/mesh/internal/types/intermediate"
+	"github.com/hashicorp/consul/internal/resource"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+func (m *Mapper) MapDestinationsToProxyStateTemplate(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) ([]controller.Request, error) {
+	var destinations pbmesh.Upstreams
+	err := res.Data.UnmarshalTo(&destinations)
+	if err != nil {
+		return nil, err
+	}
+
+	// Look up workloads for this destinations.
+	sourceProxyIDs := make(map[resource.ReferenceKey]struct{})
+	var result []controller.Request
+	for _, prefix := range destinations.Workloads.Prefixes {
+		resp, err := rt.Client.List(ctx, &pbresource.ListRequest{
```

Comment:
(later) I suspect there's something about this general shape of "using a workload selector to iterate" that could be extracted from this and the endpoints controller since you have to know in both cases to separately walk to the prefix list, and the exact list and execute two different queries